feat(ledger): support ledger device in keyring controller

zoltanarvai opened the pull request:
**Ledger Integration**
Adding support to connecting to a ledger device and retrieving a default account from the device.

**Description**

- BREAKING:
No breaking changes

- FIXED:
N/A

- CHANGED:
KeyringController.signTypedMessage -> Added support to check if the account is a ledger device and if so, propagate the signing request to the device.

- REMOVED:
N/A

- ADDED:
KeyringController.getLedgerKeyring() - Retrieve the keyring for ledger
KeyringController.connectLedgerHardware() - Ability to connect to a device using a transport
KeyringController.unlockLedgerDefaultAccount() - Retrieve the first account from the device
KeyringController.forgetLedger() - Reset the state of the keyring

- DEPRECATED:
N/A

- SECURITY:
N/A

**Checklist**

- [x] Tests are included if applicable
- [x] Any added code is fully documented

zoltanarvai commented on package.json:
```diff
@@ -38,6 +38,7 @@
     "@ethereumjs/common": "^2.3.1",
     "@ethereumjs/tx": "^3.2.1",
     "@metamask/contract-metadata": "^1.33.0",
+    "@ledgerhq/metamask-keyring": "https://github.com/LedgerHQ/metamask-keyring",
```
We'll update this in a day or two (once it's published on our side.) Until then this PR is a WIP

mcmire commented on src/keyring/KeyringController.ts:
```diff
@@ -743,6 +755,87 @@ export class KeyringController extends BaseController<
+   * @param deviceId - The device ID to connect to
+   * @returns The name of the currently open application on the device
+   */
+  async connectLedgerHardware(transport: any, deviceId: string) {
```
Is there a type we can use besides `any`? We try to avoid this if we can.

mcmire commented on src/keyring/KeyringController.ts:
```diff
@@ -442,6 +444,16 @@ export class KeyringController extends BaseController<
       const address = normalizeAddress(messageParams.from);
+
+      const ledgerKeyring = await this.getLedgerKeyring();
+      const isLegerAccount = await ledgerKeyring.managesAccount(address);
+
+      if (isLegerAccount) {
```
``` suggestion
      const isLedgerAccount = await ledgerKeyring.managesAccount(address);

      if (isLedgerAccount) {
```

zoltanarvai replied:
fixed

mcmire reviewed (commented):
Hi @zoltanarvai. Thanks for this PR. I left some feedback below.

zoltanarvai commented on src/keyring/KeyringController.test.ts:
```diff
@@ -911,4 +924,263 @@ describe('KeyringController', () => {
+        baseConfig,
+      );
+
+      await keyringController.createNewVaultAndKeychain(password);
+      ledgerKeyring = await keyringController.getLedgerKeyring();
+
+      ledgerKeyring.setApp({
+        signTransaction: sinon.stub(),
+        getAddress: sinon.stub().resolves({
+          address: '0xe908e4378431418759b4f87b4bf7966e8aaa5cf2',
+        }),
+        signEIP712HashedMessage: sinon.stub(),
+        signPersonalMessage: sinon.stub(),
+      });
+
+      await keyringController.unlockLedgerDefaultAccount();
```
Hey @mcmire ! Thanks for the feedback! This setup is needed for the Ledger Keyring only. (it's in its own describe block). The ledger nano needs to have confirmation from the user to perform actions that's what we are mocking out here. There is too steps we always need:
- Have an app with a transport set (that serves as the bridge between the keyring and the device. (setApp(..) call)
- Have one the first account unlocked so they keyring has something (unlockLedgerDefaultAccount() call)

zoltanarvai commented on src/keyring/KeyringController.test.ts:
```diff
@@ -263,6 +264,8 @@ describe('KeyringController', () => {
   });
 
   it('should throw when given invalid version', async () => {
+    await keyringController.getLedgerKeyring();
```
It's not really breaking, the main reason for that (we followed the KeyStone implementation) is during signedType data we need to see if the account belongs to the ledger keyring if so we'll have to let the device do the signing. This piece makes sure that the keyring is always added. However we can definitely be even more "backward compatible", to not do the check at all and just say if no ledgerKeyring is registered then just skip to the next block of code. I'll update accordingly.

zoltanarvai replied:
Update done

zoltanarvai commented on src/keyring/KeyringController.test.ts:
```diff
@@ -45,7 +46,7 @@ describe('KeyringController', () => {
     keyrings: Keyring[];
   };
-  const additionalKeyrings = [QRKeyring];
+  const additionalKeyrings = [QRKeyring, LedgerKeyring];
```
This is kind of a breaking change. Once we have the new controller package referenced in the mobile app the LedgerKeyring needs to be set as an additional Keyring. To be fair, this is the same requirement as the already existing QRKeyring has.

zoltanarvai commented:
> Hi @zoltanarvai. Thanks for this PR. I left some feedback below.

Thank you very much for the quick feedback, I've addressed your comments. Please let me know what you think.

gantunesr commented on src/keyring/KeyringController.ts:
```diff
@@ -18,6 +18,8 @@ import {
   MetaMaskKeyring as QRKeyring,
   IKeyringState as IQRKeyringState,
 } from '@keystonehq/metamask-airgapped-keyring';
+import Transport from '@ledgerhq/hw-transport';
```
What is `Transport`? I'm not familiar with this concept

zoltanarvai replied:
Transport is the base type for all transports with ledger. I've only added it as @mcmire preferred the transport function parameter to be typed instead of having it as `any`

gantunesr commented on src/keyring/KeyringController.ts:
```diff
@@ -442,6 +445,16 @@ export class KeyringController extends BaseController<
     try {
       const address = normalizeAddress(messageParams.from);
+
+      const ledgerKeyring = await this.getLedgerKeyring();
+      const isLedgerAccount = await ledgerKeyring.managesAccount(address);
+
+      if (isLedgerAccount) {
+        return privates
+          .get(this)
+          .keyring.signTypedMessage(messageParams, { version });
+      }
```
For a next iteration we could create methods to hold each block of code (Ledger, Keystone, etc)

gantunesr replied:
Its not necessary to address it in this PR

zoltanarvai replied:
Ideally we should push down the code into the respective keyrings as the "account check" logic is already performed by the Ethereum keyring controller. However even for the MM code the signTypedData logic lives in the app and not in the keyring.

gantunesr commented on src/keyring/KeyringController.ts:
```diff
@@ -743,6 +756,87 @@ export class KeyringController extends BaseController<
+   * @param deviceId - The device ID to connect to
+   * @returns The name of the currently open application on the device
+   */
+  async connectLedgerHardware(transport: Transport, deviceId: string) {
```
Add return type

gantunesr commented on src/keyring/KeyringController.ts:
```diff
@@ -743,6 +756,87 @@ export class KeyringController extends BaseController<
+   *
+   * @returns The default (first) account on the device
+   */
+  async unlockLedgerDefaultAccount() {
```
Add return type

zoltanarvai replied:
done

gantunesr commented on src/keyring/KeyringController.ts:
```diff
@@ -743,6 +756,87 @@ export class KeyringController extends BaseController<
+    const address = await keyring.getDefaultAccount();
+    return {
+      address,
+      balance: `0x0`,
```
Will this value be hardcode for the final version?

gantunesr reviewed (commented):
Hi @zoltanarvai, thanks for this development! it's looking good. I left some comments and question

zoltanarvai commented on tsconfig.json:
```diff
@@ -10,7 +10,8 @@
     "strict": true,
-    "target": "es6"
+    "target": "es6",
+    "skipLibCheck": true, // Skip type checking of declaration files.
```
during building the typescript compiler (while builds the package) give me a weird TS error on an internal / referenced package (ethereumjs). Indeed it's valid, the issue is in the typing of the library. Are you okay with switching this off? @gantunesr @mcmire

Gudahtt replied:
Only as a last resort. This results in any type declarations defaulting to `any` if they have errors, so it decreases type safety and can let errors sneak in without us noticing.

What was the error you encountered?

zoltanarvai replied:
Getting this:
```
node_modules/@ledgerhq/metamask-keyring/node_modules/@ethereumjs/tx/node_modules/@ethereumjs/common/dist/types.d.ts:45:6 - error TS1336: An index signature parameter type cannot be a type alias. Consider writing '[key: string]: string | AccountState' instead.

45     [key: PrefixedHexString]: PrefixedHexString | AccountState;
        ~~~
```

Gudahtt replied:
Gotcha - we have seen that error already in other repositories. I don't fully understand what's going on there, but we've found that the error goes away after a TypeScript update. I can look into updating TypeScript.

zoltanarvai replied:
That'd be great, I did try and didn't seem to resolve the issue.

Gudahtt replied:
I have just confirmed locally that updating TypeScript fixes this issue. It introduced a number of other build issues, but those were all easily resolved and will be addressed as part of the update PR.

gantunesr replied:
@zoltanarvai since you were able to solve this issue it should be safe to revert this change

gantunesr commented:
Just a note to @Gudahtt and @mcmire, this is not going to be merged into `main`, we'll be working with a feature branch for the moment ([MetaMask:feat/ledger-integration](https://github.com/MetaMask/controllers/tree/feat/ledger-integration))

gantunesr reviewed (approved):
LGTM. Pinging @mcmire @Gudahtt to verify if there is any other concern over this PR
